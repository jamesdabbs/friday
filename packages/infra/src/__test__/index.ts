import { Testing } from 'cdk8s';

export function synth(construct: any, { name = 'construct', ...params }: Record<string, unknown> = {}) {
  const app = Testing.app();
  const chart = new construct(app, name, params);
  return Testing.synth(chart);
}

export function find(chart: any[], kind: string, name?: string) {
  return chart.find(value => {
    if (value.kind !== kind) { return false }
    if (name && value.metadata?.name !== name) { return false }

    return true
  })
}
