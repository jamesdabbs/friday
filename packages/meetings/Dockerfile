ARG NEST_PLATFORM
FROM $NEST_PLATFORM as nest-platform

FROM node:16-slim AS builder
WORKDIR /app

RUN apt-get -qq update && \
  apt-get -yqq --no-install-recommends install libssl-dev && \
  rm -rf /var/lib/apt/lists/*

COPY package* ./
COPY --from=nest-platform /lib/ /lib/nest-platform

RUN npm install /lib/nest-platform
RUN --mount=type=secret,id=npmrc,dst=/app/.npmrc --mount=type=cache,target=~/.npm npm ci

COPY prisma prisma/
RUN npx prisma generate

COPY . .

ENV NODE_ENV=development
CMD ["npm", "run", "start:dev"]
