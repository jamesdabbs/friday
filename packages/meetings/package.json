{
  "name": "meetings",
  "version": "0.0.1",
  "description": "",
  "author": "James Dabbs <james.dabbs@gmail.com>",
  "private": true,
  "license": "UNLICENSED",
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "nest build",
    "start": "nest start",
    "start:dev": "nest-platform dev",
    "start:debug": "nest start --debug --watch",
    "start:prod": "node dist/main"
  },
  "dependencies": {
    "@jamesdabbs/nest-platform": "0.0.5",
    "@nestjs/common": "^7.6.16",
    "@nestjs/core": "^7.6.16",
    "@nestjs/microservices": "^7.6.17",
    "@nestjs/platform-express": "^7.6.15",
    "@prisma/client": "^2.23.0",
    "kafkajs": "^1.15.0",
    "prisma": "^2.23.0",
    "reflect-metadata": "^0.1.13",
    "rimraf": "^3.0.2",
    "rxjs": "^6.6.6"
  },
  "devDependencies": {
    "@nestjs/cli": "^7.6.0",
    "@nestjs/schematics": "^7.3.0",
    "@types/express": "^4.17.11",
    "@types/luxon": "^1.26.5",
    "@types/node": "^14.14.36",
    "run-script-webpack-plugin": "^0.0.11",
    "ts-loader": "^8.3.0",
    "ts-node": "^9.1.1",
    "tsconfig-paths": "^3.9.0",
    "typescript": "^4.2.3",
    "webpack": "^5.39.1",
    "webpack-cli": "^4.7.2",
    "webpack-node-externals": "^2.5.2"
  }
}
