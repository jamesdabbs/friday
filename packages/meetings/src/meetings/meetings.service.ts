import { Inject, Injectable } from '@nestjs/common';
import { Instance, Meeting, Person, PrismaClient } from '@prisma/client'
import { PRISMA_CLIENT } from '@jamesdabbs/nest-platform';

function inHours(n: number) {
  const date = new Date()
  date.setTime(date.getTime() + (n * 60 * 60 * 1000))
  return date
}

export type EagerLoadedInstance = Instance & { meeting: Meeting, emcee: Person, scribe: Person }

const james: Person = { id: 1, slackName: '@james' }
const meetings: Person = { id: 2, slackName: '@meetings' }

const instances: EagerLoadedInstance[] = [
  {
    id: 1,
    meetingId: 1,
    meeting: {
      id: 1,
      name: 'Tech Leads Core',
      slackChannel: '_robots'
    },
    at: inHours(6),
    emceeId: 1,
    emcee: james,
    scribeId: 2,
    scribe: meetings,
    agendaUrl: 'http://example.com'
  }
]

@Injectable()
export class MeetingsService {
  constructor(@Inject(PRISMA_CLIENT) private readonly prisma: PrismaClient) {}

  async today(): Promise<EagerLoadedInstance[]> {
    const meetings = await this.prisma.instance.findMany({
      include: { meeting: true, emcee: true, scribe: true }
    })

    console.log({ meetings })

    return instances
  }
}
