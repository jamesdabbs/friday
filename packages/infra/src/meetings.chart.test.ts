import { MeetingsChart } from './meetings.chart';
import { find, synth } from './__test__';

describe(MeetingsChart, () => {
  function mainContainer(chart: any) {
    const deploy = find(chart, 'Deployment', 'meetings')
    return deploy?.spec?.template?.spec?.containers?.find((c: any) => c.name === 'meetings')
  }

  function cronJob(chart: any) {
    return find(chart, 'CronJob', 'meetings-daily-reminder')
  }

  describe('with defaults', () => {
    const chart = synth(MeetingsChart)

    it('matches the snapshot', () => {
      expect(chart).toMatchSnapshot()
    })

    it('has a liveness probe', () => {
      const container = mainContainer(chart)

      expect(
        container.livenessProbe.httpGet
      ).toBeTruthy()
    })

    it('schedules a job', () => {
      const job = cronJob(chart)
      expect(
        job?.spec?.schedule
      ).toBeTruthy()
    })
  })

  describe('with overridden schedule', () => {
    const chart = synth(MeetingsChart, { schedule: '@hourly' })

    it('uses the provided schedule', () => {
      const job = cronJob(chart)

      expect(
        job?.spec?.schedule
      ).toEqual('@hourly')
    })
  })
});
