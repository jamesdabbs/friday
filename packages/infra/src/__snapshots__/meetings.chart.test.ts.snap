// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeetingsChart with defaults matches the snapshot 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "meetings-db",
    },
    "spec": Object {
      "externalIPs": Array [],
      "ports": Array [
        Object {
          "port": 5432,
        },
      ],
      "selector": Object {
        "cdk8s.statefulset": "construct-meetings.db.sts-c83f5e98",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "StatefulSet",
    "metadata": Object {
      "name": "meetings-db",
    },
    "spec": Object {
      "podManagementPolicy": "OrderedReady",
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.statefulset": "construct-meetings.db.sts-c83f5e98",
        },
      },
      "serviceName": "meetings-db",
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.statefulset": "construct-meetings.db.sts-c83f5e98",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "POSTGRES_PASSWORD",
                  "value": "hunter2",
                },
              ],
              "image": "postgres",
              "imagePullPolicy": "IfNotPresent",
              "name": "postgres",
              "ports": Array [
                Object {
                  "containerPort": 5432,
                },
              ],
              "volumeMounts": Array [],
            },
          ],
          "volumes": Array [],
        },
      },
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "meetings",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.deployment": "construct-meetings.deployment-c8bcd4fd",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.deployment": "construct-meetings.deployment-c8bcd4fd",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "DATABASE_URL",
                  "value": "postgres://postgres:hunter2@meetings-db:5432",
                },
              ],
              "image": "meetings",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "failureThreshold": 3,
                "httpGet": Object {
                  "path": "/health",
                  "port": 3000,
                },
              },
              "name": "meetings",
              "ports": Array [],
              "volumeMounts": Array [],
            },
          ],
          "volumes": Array [],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "meetings",
    },
    "spec": Object {
      "externalIPs": Array [],
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 3000,
        },
      ],
      "selector": Object {
        "cdk8s.deployment": "construct-meetings.deployment-c8bcd4fd",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "name": "meetings.ingress.local",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "meetings.localhost",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "meetings",
                  "servicePort": 80,
                },
              },
            ],
          },
        },
      ],
    },
  },
  Object {
    "apiVersion": "batch/v1beta1",
    "kind": "CronJob",
    "metadata": Object {
      "name": "meetings-daily-reminder",
    },
    "spec": Object {
      "jobTemplate": Object {
        "spec": Object {
          "template": Object {
            "spec": Object {
              "containers": Array [
                Object {
                  "command": Array [
                    "curl",
                    "-XPOST",
                    "meetings/sendDailyReminders",
                  ],
                  "image": "curlimages/curl",
                  "imagePullPolicy": "IfNotPresent",
                  "name": "daily-reminder",
                },
              ],
              "restartPolicy": "Never",
            },
          },
          "ttlSecondsAfterFinished": 60,
        },
      },
      "schedule": "0 9 * * *",
    },
  },
]
`;
